### ISSUE TITLE
We're working on a new milter program to replace iRedAPD

### ISSUE THREAD

**Issue by iredmail** [iredmail/iRedAPD#11]
Dear all,

Just want to share with you that we're developing a new milter program to replace iRedAPD, so don't feel sad if no much commits here.

### Why?

iRedAPD doesn't get full email message (headers + body), hence it cannot do advanced access control. But milter can do this.

### More details

- The milter program is written in Golang programming language Ideally it should run faster than iRedAPD.
- All iRedAPD plugins will be implemented in first stable release (v1.0).
- Source code will be published when v1.0 is out, all further development will be open source too.

If you need some new features which are not yet implemented in iRedAPD, reply below to let us know. :)

**Comment by gizahNL** [comment]
Feature requests:
-Full DKIM/SPF
-Compliant ARC support
-DMARC: handle sending of dmarc responses
-Proper functioning of throttling & such when combined with Dovecot Submission (needed for BURL & such) (see: https://doc.dovecot.org/admin_manual/submission_server/)

**Comment by iredmail** [comment]
Thanks @gizahNL.

- First stable release of this milter program will just offer same features as iRedAPD.
- DKIM signing / verification, SPF verification, ARC support, DMARC, all these are on our todo list, but each may take some time.

**Comment by iredmail** [comment]
Request from [iRedMail forum](https://forum.iredmail.org/topic19253.html): Add Warning/Disclaimer to all email from external domains

**Comment by iredmail** [comment]
Block clients based on country (GeoIP).

**Comment by kprkpr** [comment]
I don't know if milter is responsible to things like replacing body or adding disclaimers, because having any alternative a bit modern to altermime, witch can add better the disclaimers or edit content... can be great (For example, companies wanting that below signature has to be any image and not below all message, and so on..) altermime at basic funcionality works but I think is somewhat basic

**Comment by iredmail** [comment]
> I don't know if milter is responsible to things like replacing body or adding disclaimers

Yes this is what milter can do.

**Comment by thinkwelltwd** [comment]
Could the database primary keys be UUID, or some other machine-specific unique value such as [snowflakeid](https://github.com/godruoyi/go-snowflake)? This would help make high-availability setups more achievable.

**Comment by iredmail** [comment]
SQL high-availability cluster should avoid this issue internally. :)

**Comment by thinkwelltwd** [comment]
> SQL high-availability (multi-master) cluster should avoid this issue internally. :)

Yes, PKs are indeed internal, and Multi Master is an internal detail. 

Designing the new database schema with HA in mind is what I'm driving at. Sequences are one of the hard problems with HA and force database replication to be synchronous. Avoiding sequences will make HA / MultiMaster options much more flexible.

**Comment by iredmail** [comment]
MySQL doesn't have column type "uuid", if we store uuid as string in MySQL, it's much slower than BIGINT (as primary key).

**Comment by thinkwelltwd** [comment]
I'm not advocating for UUID specifically; but rather a non-sequential value with uniqueness properties that make it suitable for generating values on multiple systems.

I'd be well satisfied with snowflakeid, as I'd mentioned in my original post. snowflakeid is stored as bigint, and could be generated in application code, or as a database function.

**Comment by iredmail** [comment]
Feature request: Add a warning text at the beginning of message body if email is sent from external user.

**Comment by iredmail** [comment]
Feature request: SRS.

**Comment by donatid** [comment]
Hello,

Do you have updates and the release schedule on this project?

Thank you

**Comment by iredmail** [comment]
- Our focus was shifted to our new product (iRedMail Pro[1]) when the milter development was about 80% done, now the new product is 70% done.
- We will work out a milter program without any iRedAPD plugins first in about 2 months - since iRedAPD works well and no rush to replace it.

[1] https://docs.iredmail.org/pro.html  [iRedMail Pro will combine all features offered by classical shell-based iRedMail installer, iRedMail Easy platform, and iRedAdmin-Pro.]

**Comment by donatid** [comment]
Thanks for the explanation. I can't wait for the new milter apps.

**Comment by git001** [comment]
Sounds really great. I also switched to golang for some tools :smile:  https://github.com/git001/caddyv2-upload .  
Happy to help if needed

Let me add a plugin request here for the future:  
MTA-STS Resolver something like https://github.com/Snawoot/postfix-mta-sts-resolver

**Comment by l00v3** [comment]
Hello, wandering what is the timeline of the SRS milter program. We have another problem. With SRS checks enabled, we get SPF fails in spamassassin:
FORGED_SPF_HELO=1
SPF_FAIL=0.001 (manually lowered the score)

But it should pass:
Received-SPF: Pass (protection.outlook.com: domain of domain.si designates
    195.245.xxx.xxx as permitted sender) receiver=protection.outlook.com;
    client-ip=195.245.xxx.xxx; helo=mail.domain.si; pr=C

Quite a lot of legal messages are going into SPAM, and customers are getting annoyed. Because we have a lot of forwarding emails, we cannot disable SRS.
Thank you for checking this one.
T

**Comment by l00v3** [comment]
Hello @iredmail, are there any news on this?

**Comment by iredmail** [comment]
Preparing to release milter with just one plugin "reject_forged_sender" in first stable version. 

Since iRedAPD is working fine, we tend to replace it slowly to save our time.

let me know if you want to help test it.

**Comment by l00v3** [comment]
I don't want to be rude or anything, and I know you have a lot of work, but shouldn't this be a quite high priority? And yes, I can help test it.
Also we have [@249](https://github.com/iredmail/iRedMail/issues/249) issue since the beginning.

**Comment by iredmail** [comment]
We're busy working on the iRedMail Enterprise Edition (EE), and v1.0-beta3 is out today. You can give it a try if you're interested.
https://docs.iredmail.org/install.iredmail.enterprise.html

iRedMail EE combines iRedMail installer + iRedMail Easy platform + iRedAdmin-Pro in one single binary, it's the highest priority.

We will try to finish this milter after iRedMail EE 1.0 stable release. Sorry about this.

**Comment by jfwelstead** [comment]
Is the new Milter to replace iredAPD on track for release in 2024

**Comment by git001** [comment]
As there is now EE version 1.0.1 out, what's the plan for the iRedAPD ?

**Comment by iredmail** [comment]
> As there is now EE version 1.0.1 out, what's the plan for the iRedAPD ?

EE will use iRedAPD for some more time, until milter is out and replaces all iredapd features.
Eventually iredapd will be dropped and milter is the future.
